Allow Base64Decoder to ignore space chars, add IsValid methods and tests

heathbm opened the pull request:
Implementation of api-approved issue: https://github.com/dotnet/runtime/issues/76020  

### Goals regarding the changes to decoding:

1. Chars now ignored by Decoding methods:
    - 9: Line feed
    - 10: Horizontal tab
    - 13: Carriage return
    - 32: Space
    - *Vertical tab omitted*

2. Performance:
    - The best-case path (when no chars to be ignored are encountered) should remain as untouched as possible.
    - The worst-case path: should still perform at O(1n)

3. The `Base64` decoding methods should handle whitespace chars as the `Convert.ToBase64` decoding methods. A large number of tests have been carried out to ensure this is the case: https://gist.github.com/heathbm/f59662bd2334761d28288755a34e29ec

4. Existing tests should not need to be altered. At the moment, a slight modification has been made to `Base64DecoderUnitTests.cs` due to the fact that invalid ranges cannot as easily be inferred with the length of the input span. If this is a deal-breaker, a change can be made, that would involve the decoding loop to look at the next block before writing the current one.

heathbm committed:
Allow Base64Decoder to ignore space chars, add IsValid methods and tests

Chars now ignored:
9: Line feed
10: Horizontal tab
13: Carriage return
32: Space
-- Vertical tab omitted

heathbm commented:
This test:   https://github.com/dotnet/runtime/blob/0c4ee9e749167a2185f59b4a16c8a5c62e0ec5df/src/libraries/System.Security.Cryptography/tests/Base64TransformsTests.cs#L304  
is no longer throwing an exception: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-79334-merge-b56bfafc0ea743d08d/System.Security.Cryptography.Tests/1/console.b696fe8a.log?helixlogtype=result  

This seems reasonable.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,214 @@
+            }
+        }
+
+        private static IndexOfAnyValues<byte> ValidBase64BytesSortedAsc => IndexOfAnyValues.Create(new byte[] {
+            9, 10, 13, 32, 43, 47,                   //Line feed, Horizontal tab, Carriage return, Space, +, /
+            48, 49, 50, 51, 52, 53, 54, 55, 56, 57,  //0..9,
+            61,                                      //=
+            65, 66, 67, 68, 69, 70, 71, 72,          //A..H
+            73, 74, 75, 76, 77, 78, 79, 80,          //I..P
+            81, 82, 83, 84, 85, 86, 87, 88,          //Q..X
+            89, 90, 97, 98, 99, 100, 101, 102,       //Y..Z, a..f
+            103, 104, 105, 106, 107, 108, 109, 110,  //g..n
+            111, 112, 113, 114, 115, 116, 117, 118,  //o..v
+            119, 120, 121, 122,                      //w..z
+        });
```
These should be written as
```c#
private static readonly IndexOfAnyValues<byte> s_validBase64Bytes =
    IndexOfAnyValues.Create("abcde..."u8);
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,214 @@
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <param name="decodedLength">The maximum length (in bytes) if you were to decode the base 64 encoded text <paramref name="base64TextUtf8"/> within a byte span.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static unsafe bool IsValid(ReadOnlySpan<byte> base64TextUtf8, out int decodedLength)
```
The common case we want to optimize for is when there are no whitespace/invalid characters present.

The current implementation is doing two passes over the input. You can instead write it along the lines of
```c#
int indexOfPaddingInvalidOrWhitespace = base64Text.IndexOfAnyExcept(s_validBase64Chars);

int length;
if ((uint)indexOfPaddingInvalidOrWhitespace >= (uint)base64Text.Length)
{
    length = base64Text.Length;
}
else
{
    char c = base64Text[indexOfPaddingInvalidOrWhitespace];
    if (c != EncodingPad)
    {
        // ... count whitespace / check for invalid characters
    }

    // Check that the remaining chars are only '=' and whitespace
    // Check that the number of padding chars makes sense
}

decodedLength = length...

return (length & 3) == 0;
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,214 @@
+            }
+
+            // Check for invalid chars
+            int indexOfFirstNonBase64 = base64TextUtf8.IndexOfAnyExcept(ValidBase64BytesSortedAsc);
```
This shouldn't be consuming `=` characters, something like `"a=b="` isn't valid input.

We are likely missing test coverage here as well, e.g.
- padding in the middle of the input
- more padding than needed
- whitespace in between padding at the end

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,214 @@
+            // Check if there are chars that need to be ignored while determining the length
+            if (base64TextUtf8.IndexOfAny(BytesToIgnore) > -1)
+            {
+                fixed (byte* srcBytes = &MemoryMarshal.GetReference(base64TextUtf8))
```
There should be no need to switch to unsafe code here, a simple `foreach` over the span will be just as fast.

MihaZupan reviewed (commented):
Thanks for looking into this again!

I only looked through the `IsValid` part.

cc: @gfoidl

heathbm committed:
Address PR feedback regarding Base64.IsValid

heathbm commented:
Thank you for the feedback @MihaZupan I believe I have addressed those comments in my latest commit.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,216 @@
+            }
+
+            // Remove padding to get exact length
+            decodedLength = (length / 4 * 3) - paddingCount;
```
```suggestion
            decodedLength = (int)((uint)length / 4 * 3) - paddingCount;
```
Ugly, but [better codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAWQAMACkHCANjAEBzDAAsAlAwC8APgZzFKhgHoGKBgCoGdANxce9BtNHkpQ3fKVrNOlxlUSJAK7SqnoeRibqFtY0AL5AA).

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,216 @@
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsCharToBeIgnored(char aChar)
```
See above.
Or will the compilers be able to optimize this (in the near future)?

gfoidl replied:
If this method takes an `int` as argument, then it could be used here and in the decoder, so moved to a shared place, thus avoiding the duplication.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,216 @@
+                }
+            }
+
+            if (length % 4 != 0)
```
TIL: here is no need for the uint-cast, the JIT will emit goog code 👍🏻.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,216 @@
+            }
+
+            // Remove padding to get exact length
+            decodedLength = (length / 4 * 3) - paddingCount;
```
```suggestion
            decodedLength = (int)((uint)length / 4 * 3) - paddingCount;
```

gfoidl reviewed (commented):
Had a quick look over the code. I'd like to have one point addressed, then will do a more thorough review.

---

When vectorized code detects an invalid char (could be whitespace), then it falls back to pure scalar processing, thus leaving perf on the table.
Often whitespace is inserted after 76 chars, so we should try to optimize that case.

Take that example:
```c#
Span<byte> bytes = new byte[1000];
Random.Shared.NextBytes(bytes);

string base64Text = Convert.ToBase64String(bytes, Base64FormattingOptions.InsertLineBreaks);
ReadOnlySpan<byte> base64 = Encoding.ASCII.GetBytes(base64Text);

OperationStatus status = Base64.DecodeFromUtf8(base64, bytes, out int consumed, out int written);
```
Here line-breaks are inserted every 76-chars. So after decoding the first "set", all the remainder is done scalar, whilst it could be done vectorized too. This is what I meant in https://github.com/dotnet/runtime/pull/78951#discussion_r1034637033.

Do to this you can use something like:
```c#
static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
{
    OperationStatus status;
    int consumed = 0;
    int written = 0;
    int totalConsumed = 0;
    int totalWritten = 0;

    while (true)
    {
        // DecodeFromUtf8Core is the current implementation from main, just renamed and made private
        status = DecodeFromUtf8Core(utf8, bytes, out consumed, out written, isFinalBlock);
        totalConsumed += consumed;
        totalWritten += written;

        if (status != OperationStatus.InvalidData)
        {
            break;
        }

        // Found invalid data, check if it's whitespace and can be skipped
        utf8 = utf8.Slice(consumed);
        bytes = bytes.Slice(written);

        if (!TrySkipWhitespace(utf8, out consumed))
        {
            break;
        }

        utf8 = utf8.Slice(consumed);
    }

    bytesConsumed = totalConsumed;
    bytesWritten = totalWritten;
    return status;
}

// This scans potentially to the end. There could be a more robust approach.
// Maybe try IndexOfAnyValues here?
static bool TrySkipWhitespace(ReadOnlySpan<byte> utf8, out int consumed)
{
    for (int i = 0; i < utf8.Length; ++i)
    {
        if (!IsByteToBeIgnored(utf8[i]))
        {
            consumed = i;
            return true;
        }
    }

    consumed = 0;
    return false;
}
```
I hope you get the idea. 
Note: the shown code doesn't handle edge-cases, etc. so that should be considered, also if true invalid data is present that it's not stuck in an endless loop.

heathbm committed:
Address PR feedback: General optimizations

heathbm committed:
Address PR feedback: Use vectorized decoding while enough src

heathbm commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -84,142 +87,199 @@ public static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Spa
-                    WriteThreeLowOrderBytes(dest, i0);
-                    dest += 3;
+                    int remainingBytes = (int)(srcEnd - src);
+                    if (remainingBytes > 0 && remainingBytes < 4)
```
I still need to check for 0 with a uint, is this still worth the cast?

heathbm commented:
> Had a quick look over the code. I'd like to have one point addressed, then will do a more thorough review.
> 
> When vectorized code detects an invalid char (could be whitespace), then it falls back to pure scalar processing, thus leaving perf on the table. Often whitespace is inserted after 76 chars, so we should try to optimize that case.
> 
> Take that example:
> 
> ```cs
> Span<byte> bytes = new byte[1000];
> Random.Shared.NextBytes(bytes);
> 
> string base64Text = Convert.ToBase64String(bytes, Base64FormattingOptions.InsertLineBreaks);
> ReadOnlySpan<byte> base64 = Encoding.ASCII.GetBytes(base64Text);
> 
> OperationStatus status = Base64.DecodeFromUtf8(base64, bytes, out int consumed, out int written);
> ```
> 
> Here line-breaks are inserted every 76-chars. So after decoding the first "set", all the remainder is done scalar, whilst it could be done vectorized too. This is what I meant in [#78951 (comment)](https://github.com/dotnet/runtime/pull/78951#discussion_r1034637033).
> 
> Do to this you can use something like:
> 
> ```cs
> static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
> {
>     OperationStatus status;
>     int consumed = 0;
>     int written = 0;
>     int totalConsumed = 0;
>     int totalWritten = 0;
> 
>     while (true)
>     {
>         // DecodeFromUtf8Core is the current implementation from main, just renamed and made private
>         status = DecodeFromUtf8Core(utf8, bytes, out consumed, out written, isFinalBlock);
>         totalConsumed += consumed;
>         totalWritten += written;
> 
>         if (status != OperationStatus.InvalidData)
>         {
>             break;
>         }
> 
>         // Found invalid data, check if it's whitespace and can be skipped
>         utf8 = utf8.Slice(consumed);
>         bytes = bytes.Slice(written);
> 
>         if (!TrySkipWhitespace(utf8, out consumed))
>         {
>             break;
>         }
> 
>         utf8 = utf8.Slice(consumed);
>     }
> 
>     bytesConsumed = totalConsumed;
>     bytesWritten = totalWritten;
>     return status;
> }
> 
> // This scans potentially to the end. There could be a more robust approach.
> // Maybe try IndexOfAnyValues here?
> static bool TrySkipWhitespace(ReadOnlySpan<byte> utf8, out int consumed)
> {
>     for (int i = 0; i < utf8.Length; ++i)
>     {
>         if (!IsByteToBeIgnored(utf8[i]))
>         {
>             consumed = i;
>             return true;
>         }
>     }
> 
>     consumed = 0;
>     return false;
> }
> ```
> 
> I hope you get the idea. Note: the shown code doesn't handle edge-cases, etc. so that should be considered, also if true invalid data is present that it's not stuck in an endless loop.

@gfoidl Thank you very much for the example. I had this idea with the `IsValid` method too, where I sliced until I hit padding, then proceeded to loop over the remaining bytes. 

However, with the decode method, one of the goals is "The Base64 decoding methods should handle whitespace chars as the `Convert.ToBase64`". To elaborate, Convert can handle any number of whitespace in any position.

So your suggestion will not work:

> ```cs
> ...
> if (!TrySkipWhitespace(utf8, out consumed))
> {
>     break;
> }
> 
> utf8 = utf8.Slice(consumed);
> ...

As this would fail for example: " Y Q = = ". Yet Convert would be able to parse it.
As a compromise, if invalid bytes are hit while decoding with the vectorized paths, block by block decoding will take over, until invalid bytes have been skipped. Then, vectorized decoding will try to resume if possible.

I have implemented or addressed your other comments also.

I hope my approach is not too 'bloaty'. "The Base64 decoding methods should handle whitespace chars as the `Convert.ToBase64`" + "Existing tests should not need to be altered" resulted in more code than I initially expected. 
I would also like to reiterate, if a complete PR is submitted while this one is still open, I would be happy to close this one.

gfoidl commented:
> So your suggestion will not work:

Well it works...at least under the presupposition that this is an suggestion to get you on the right road 😉

> invalid bytes are hit while decoding with the vectorized paths, block by block decoding will take over, until invalid bytes have been skipped. Then, vectorized decoding will try to resume if possible.

Just to specify: only whitespace is skipped, not any invalid bytes.
(I know you know this, but as written it can be mis-read)

>  Convert can handle any number of whitespace in any position.
> ...
> " Y Q = = ". Yet Convert would be able to parse it.

Yep, but that's an artificial example which (I assume) won't be quite often in real-world base64 encoded bytes.
Basically there will be two major sets of base64 encoded bytes:
* no whitespace at all
* line-breaks after 76 chars

So I'd be happy when we can optimize for these cases. All other cases, like the one given in the quote, it should work, but perf-wise that can be penalized (as being very uncommon).

In my suggestion `TrySkipWhitespace` needs to be replaced with a correct / better implementation that
* skips whitespace and moves the pointer forward in the input-data
* signals "end" when truly invalid data is detected

That's not done in the suggestion (for simplicity, and just to outline how it could work).

Thus -- at least I hope so -- the actual code change is quite minimal. Have the workhorse-method as is and just renamed to _Core_-suffix. The call that workhorse method from the "driver"-method, that handles the checks as given in the two bullets above.

So for the case "no whitespace at all" there's only one method call more than the actual code has. So perf should be on-par.
For the case "line-breaks after 76 chars" 76 chars can be processed fast, then the line-break is skipped, another 76 chars are processed, and so on.

The example " Y Q = = " should work too, but you're rigth that this is a case I didn't consider in my suggestion so far. So lets extend that suggestion (point 4, last ->)
1. Core called with " Y Q = = " -> returns `OperationStatus.InvalidData`
2. check if truly invalid or whitespace -> whitespace, so skip it -> it remains "Y Q = = "
3. Core called with "Y Q = = " -> returns `OperationStatus.InvalidData`
4. check if truly invalid or whitspace -> first char "Y" is valid -> fallback to slow method

The "slow method" creates a buffer of block-size (e.g. `stackalloc byte[4]`) and fills that with the valid non-whitespace bytes. Except when invalid byte found, then stop. For the example that buffer would be filled with `YQ==`.
After filling that block-buffer Core is called to do actual decoding. The buffer is valid, has padding so it should be the final "block" to process. After that only only byte from the input remains, so check if that's whitespace (-> OK) or invalid data, where in this case invalid data is anything else than whitspace (-> KO).

This also means that once such a sequence -- invalid data, where the first byte is actual a valid base64 -- is detected it will be pure scalar processing, w/o going back to vectorized again. I'd take this, as
* it's an unlikey input -- constructed artificially for tests
* it keeps the code stream-lined, so the common inputs aren't penalized
* the code-change is also streamlined
* as whitespace wasn't allowed before it's not a perf-regression and that like, simply as that nothing to compare with
* bring a PR in that enabled whitespace, optimize it later with another PR if needed

> I would also like to reiterate, if a complete PR is submitted while this one is still open, I would be happy to close this one.

I don't follow. Let's bring your / this PR to a successful end.

heathbm committed:
Address PR feedback: General optimization

heathbm commented:
> > So your suggestion will not work:
> 
> Well it works...at least under the presupposition that this is an suggestion to get you on the right road 😉
> 
> > invalid bytes are hit while decoding with the vectorized paths, block by block decoding will take over, until invalid bytes have been skipped. Then, vectorized decoding will try to resume if possible.
> 
> Just to specify: only whitespace is skipped, not any invalid bytes. (I know you know this, but as written it can be mis-read)
> 
> > Convert can handle any number of whitespace in any position.
> > ...
> > " Y Q = = ". Yet Convert would be able to parse it.
> 
> Yep, but that's an artificial example which (I assume) won't be quite often in real-world base64 encoded bytes. Basically there will be two major sets of base64 encoded bytes:
> 
> * no whitespace at all
> * line-breaks after 76 chars
> 
> So I'd be happy when we can optimize for these cases. All other cases, like the one given in the quote, it should work, but perf-wise that can be penalized (as being very uncommon).
> 
> In my suggestion `TrySkipWhitespace` needs to be replaced with a correct / better implementation that
> 
> * skips whitespace and moves the pointer forward in the input-data
> * signals "end" when truly invalid data is detected
> 
> That's not done in the suggestion (for simplicity, and just to outline how it could work).
> 
> Thus -- at least I hope so -- the actual code change is quite minimal. Have the workhorse-method as is and just renamed to _Core_-suffix. The call that workhorse method from the "driver"-method, that handles the checks as given in the two bullets above.
> 
> So for the case "no whitespace at all" there's only one method call more than the actual code has. So perf should be on-par. For the case "line-breaks after 76 chars" 76 chars can be processed fast, then the line-break is skipped, another 76 chars are processed, and so on.
> 
> The example " Y Q = = " should work too, but you're rigth that this is a case I didn't consider in my suggestion so far. So lets extend that suggestion (point 4, last ->)
> 
> 1. Core called with " Y Q = = " -> returns `OperationStatus.InvalidData`
> 2. check if truly invalid or whitespace -> whitespace, so skip it -> it remains "Y Q = = "
> 3. Core called with "Y Q = = " -> returns `OperationStatus.InvalidData`
> 4. check if truly invalid or whitspace -> first char "Y" is valid -> fallback to slow method
> 
> The "slow method" creates a buffer of block-size (e.g. `stackalloc byte[4]`) and fills that with the valid non-whitespace bytes. Except when invalid byte found, then stop. For the example that buffer would be filled with `YQ==`. After filling that block-buffer Core is called to do actual decoding. The buffer is valid, has padding so it should be the final "block" to process. After that only only byte from the input remains, so check if that's whitespace (-> OK) or invalid data, where in this case invalid data is anything else than whitspace (-> KO).
> 
> This also means that once such a sequence -- invalid data, where the first byte is actual a valid base64 -- is detected it will be pure scalar processing, w/o going back to vectorized again. I'd take this, as
> 
> * it's an unlikey input -- constructed artificially for tests
> * it keeps the code stream-lined, so the common inputs aren't penalized
> * the code-change is also streamlined
> * as whitespace wasn't allowed before it's not a perf-regression and that like, simply as that nothing to compare with
> * bring a PR in that enabled whitespace, optimize it later with another PR if needed
> 
> > I would also like to reiterate, if a complete PR is submitted while this one is still open, I would be happy to close this one.
> 
> I don't follow. Let's bring your / this PR to a successful end.

Thank you for the details, I'm 100% onboard with those 3 scenarios. I will update the PR accordingly.

gfoidl commented:
Side note: please don't do the full-quotes -- that content is in the comment history anyway.

heathbm committed:
Address PR feedback: Optimize for whitespace (\r\n) every 76 bytes

heathbm commented:
@gfoidl Regarding:

> For the case "line-breaks after 76 chars" 76 chars can be processed fast, then the line-break is skipped, another 76 chars are       > processed, and so on.". 

and 

> In my suggestion TrySkipWhitespace needs to be replaced with a correct / better implementation that
> - skips whitespace and moves the pointer forward in the input-data
> - signals "end" when truly invalid data is detected

Suggested code for reference: 

````C#
...
if (!TrySkipWhitespace(utf8, out consumed))
{
    break;
}

utf8 = utf8.Slice(consumed);
...

...
static bool TrySkipWhitespace(ReadOnlySpan<byte> utf8, out int consumed)
{
    for (int i = 0; i < utf8.Length; ++i)
    {
        if (!IsByteToBeIgnored(utf8[i]))
        {
            consumed = i;
            return true;
        }
    }

    consumed = 0;
    return false;
}
...
````
`TrySkipWhitespace` skips valid bytes until whitespace is found. Those valid bytes need to be decoded also. I did not see a straightforward way to fill in the gaps with this suggestion, since those skipped bytes would be lost after the slice. However, I would like to raise another similar approach:

We can avoid scanning, since we should be able to check for the 2 whitespace chars (\r\n) after every 76 bytes. I currently do this in: https://github.com/dotnet/runtime/blob/f389289d0849a528e480efba3a96faa14f16c362/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L559  

I'd also like to annotate the vectorized code path: https://github.com/dotnet/runtime/blob/f389289d0849a528e480efba3a96faa14f16c362/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L72 

```C#
if (Avx2.IsSupported)
{
    while (end >= src)
    {
        // ---------- Scenario 1: no whitespace, will complete and exit
        bool isComplete = Avx2Decode(ref src, ref dest, end, maxSrcLength, destLength, srcBytes, destBytes);
        if (isComplete)
        {
            break;
        }

        // ---------- Scenario 2: whitespace after every 76 chars, method above will eventually fail, here whitespace is omitted in the next vectorized decode, then the path jumps to the start of the loop to repeat the process.
        isComplete = TryDecodeCurrentGroupIfWhitespaceIsSeparatingValidBytesInCommonLocation(utf8, ref src, ref dest, ref end, destLength, srcBytes, destBytes, 32);
        if (isComplete)
        {
            continue;
        }

        // ---------- Scenario 3: whitespace is somewhere else or invalid bytes are present, as soon as whitespace has been skipped, this path jumps to the start of the loop to reattempt vectorized decoding.
        // Process 4 bytes, until first set of invalid bytes are skipped.
        lastBlockStatus = IgnoreWhitespaceAndConsumeValidBytesUntilInvalidBlock(ref src, srcEnd, ref dest, destEnd, ref decodingMap, isFinalBlock, lastBlockStatus, true, out pendingSrcIncrement);
        if (lastBlockStatus != OperationStatus.Done)
        {
            break;
        }
    }
}
```

gfoidl commented:
> Suggested code for reference:

It's not and was not meant as "suggested code" which can be used 1:1. It's a piece of code to lay you the road on how it could be done. Thus I don't understand

> `TrySkipWhitespace` skips valid bytes until whitespace is found. Those valid bytes need to be decoded also.

as we had this already before. In the code fragment that I've shown above some parts were missing, which should be worked out.

For me your code with `TryDecodeCurrentGroupIfWhitespaceIsSeparatingValidBytesInCommonLocation` and the `% 76` looks too complicated. 

In my Base64-project I did a quick [stab](https://github.com/gfoidl/Base64/pull/184) based on the idea outlined above. The code for the actual decoding isn't touched -- similar to the outline (not suggestion) in https://github.com/dotnet/runtime/pull/79334#pullrequestreview-1208362190. All work is done [on top of the core decoding method](https://github.com/gfoidl/Base64/blob/b6e941ed496bd8ceff9f75237cadf16baa5e638b/source/gfoidl.Base64/Internal/Base64Impl.DecodeWhitespace.cs#L34-L187).
The basic part is [this loop](https://github.com/gfoidl/Base64/blob/b6e941ed496bd8ceff9f75237cadf16baa5e638b/source/gfoidl.Base64/Internal/Base64Impl.DecodeWhitespace.cs#L38-L67), which tries to decode as much as possible. If invalid data is found -- to which whitespace is accounted at that point -- it's tried to skip possible whitespace, then loop over. So cases with regular whitespace-patterns (e.g. linebreak after 76 chars) are handled fast and vectorized.

If that doesn't work it's fallen back to a [slow](https://github.com/gfoidl/Base64/blob/b6e941ed496bd8ceff9f75237cadf16baa5e638b/source/gfoidl.Base64/Internal/Base64Impl.DecodeWhitespace.cs#L133-L184) method, the decodes by blocks of four. Perf isn't great here, but when we land here then the input is degenerated (in my opinion), so it's not worth to have super fast perf as long as correct decoding can be done.

Note: at the moment there are some [basic tests](https://github.com/gfoidl/Base64/blob/b6e941ed496bd8ceff9f75237cadf16baa5e638b/tests/gfoidl.Base64.Tests/Internal/Base64ImplTests/Decode/Whitespace.cs) and a fuzz-run. I don't know if my PR for the change in that repo is good enough to merge or not -- anyway that's independent from this PR here.

heathbm commented:
I just checked out the code in your project. I stepped through the code and it appears that my previous commit essentially does the same as yours:
- Vectorized decode, until whitespace is hit
- Decode 4 bytes blocks until whitespace is hit, then try to go back to vectorized decoding. My method advances the pointer, yours slices from a worker method.

````C#
if (Avx2.IsSupported)
{
    while (end >= src)
    {
        // ---------- Step 1: Decode fast, until whitespace is hit
        bool isComplete = Avx2Decode(ref src, ref dest, end, maxSrcLength, destLength, srcBytes, destBytes);
        if (isComplete)
        {
            break;
        }

        // ---------- Step 2: Process 4 bytes, until first set of invalid bytes are skipped. Then go back to fast decoding.
        lastBlockStatus = IgnoreWhitespaceAndConsumeValidBytesUntilInvalidBlock(ref src, srcEnd, ref dest, destEnd, ref decodingMap, isFinalBlock, lastBlockStatus, true, out pendingSrcIncrement);
        if (lastBlockStatus != OperationStatus.Done)
        {
            break;
        }
    }
}
````

You mentioned this needed to be optimized which is why I added Step 2 bellow:

````C#
if (Avx2.IsSupported)
{
    while (end >= src)
    {
        // ---------- Step 1: no whitespace, will complete and exit
        bool isComplete = Avx2Decode(ref src, ref dest, end, maxSrcLength, destLength, srcBytes, destBytes);
        if (isComplete)
        {
            break;
        }

        // ---------- Step 2: whitespace after every 76 chars, method above will eventually fail, here whitespace is omitted in the next vectorized decode, then the path jumps to the start of the loop to repeat the process.
        isComplete = TryDecodeCurrentGroupIfWhitespaceIsSeparatingValidBytesInCommonLocation(utf8, ref src, ref dest, ref end, destLength, srcBytes, destBytes, 32);
        if (isComplete)
        {
            continue;
        }

        // ---------- Step 3: whitespace is somewhere else or invalid bytes are present, as soon as whitespace has been skipped, this path jumps to the start of the loop to reattempt vectorized decoding.
        // Process 4 bytes, until first set of invalid bytes are skipped.
        lastBlockStatus = IgnoreWhitespaceAndConsumeValidBytesUntilInvalidBlock(ref src, srcEnd, ref dest, destEnd, ref decodingMap, isFinalBlock, lastBlockStatus, true, out pendingSrcIncrement);
        if (lastBlockStatus != OperationStatus.Done)
        {
            break;
        }
    }
}
````

Step 2 here attempts to get rid of the whitespace (without scanning as we should be able to quickly detect if this is in fact the whitespace every 76 chars scenario) and go straight back to vector decoding instead of 4 byte blocks. Is this optimization not necessary?

heathbm commented:
To add a little more context as to why my PR appears quite large, a lot of my approach was also influenced by the requirement: "The Base64 decoding methods should handle whitespace chars as the Convert.ToBase64 decoding methods".  e.g. If Base64.IsValid == true, then Base64.Decode should return OperationStatus.Done.
Currently, the code in your project returns StatusOperation.InvalidData for " ", where as Convert.FromBase64String just returns an empty array. I have been using https://gist.github.com/heathbm/f59662bd2334761d28288755a34e29ec to test many scenarios. Currently, 397,488 should be passing. However, with your code, 133,538 pass and 263,902 fail. I believe the intention should be to have all these pass, such that, anything that Convert.ToBase64 can decode Base64.Decode should also be able to decode.

gfoidl commented:
> Step 2 here attempts to get rid of the whitespace (without scanning as we should be able to quickly detect if this is in fact the whitespace every 76 chars scenario) and go straight back to vector decoding instead of 4 byte blocks. 

IMO an explicit step 2 isn't needed. If invalid data is detected, check if it's whitespace and skip it. Then call the decoding-implementation again -- which will use vectorization if the source is long enough.
Only if non-whitespace is detected or the current-main-decoding implementation can't handle it, then it's fallen back to block-of-4-decoding -- which is slow, but only needed for degenerated inputs.

I'm leaning towards a KISS solution, that's only optimized for the most common cases. 
For this there's no need to have `% 76` in it. What if the line-break isn't at 76, rather at 72, 68 or that like? 76 is the default. Or what if there's a line-break and additional empty line inserted?
So I'd just slice off the whitespace that is found, then call the decoding implementation again. The code-change is way less and re-uses existing building blocks.

> The Base64 decoding methods should handle whitespace chars as the Convert.ToBase64 decoding methods

Sure, both should have the same behavior.
And in the long-term the implemenations should be unified -- so that not two implementations "of the same piece" are present.

> Currently, the code in your project returns ...

Yup...I know that my code isn't really ready now, as it's hacked out yesterday evening with only a basic test.
So thanks for the gist, I'll incorporate this.

heathbm commented:
> I'm leaning towards a KISS solution, that's only optimized for the most common cases.
> For this there's no need to have % 76 in it.

I'm definitely onboard with this, I had mixed feelings about the `% 76` code too, I will revert that change. Apologies for the confusion.

> So I'd just slice off the whitespace that is found, then call the decoding implementation again. The code-change is way less 
> and re-uses existing building blocks.

This is already being done with `IgnoreWhitespaceAndConsumeValidBytesUntilInvalidBlock` and `exitAfterFirstSkippedBytes`. However, I will update the approach so that the original `DecodeFromUtf8` is as untouched as possible.

heathbm committed:
Address PR feedback: Implement driver/worker pattern with original code

heathbm commented:
@gfoidl Feedback has been addressed regarding maintaining untouched core methods, to preserve existing code paths.

gfoidl reviewed (commented):
TBH I have some troubles to follow along the code w/o debugging it.
It still seems that it is a bit over-engineered, the [actual change](https://github.com/gfoidl/Base64/blob/13711c730adf409ef35f3459dc4bdac47bae27fd/source/gfoidl.Base64/Internal/Base64Impl.DecodeWhitespace.cs#L25-L197) could be quite straightforward.

Sure there are different approaches to the same problem, and I'm not going to say that my appraoch is better, but at least for me it's easier to read (yeah, I've written it, so it would be very strange otherwise 😉).

Please try to simplify the (intermediate) steps, as basically it's just:
```mermaid
flowchart
    B{{Start}} --> Decode
    Decode --> status{OK}
    status -->|yes| C{{Done}}
    status -->|no| BA[Try skip Whitespace] --> status1{OK}
    status1 -->|yes| Decode
    
    status1 -->|no| BC[Fill block-buffer w/o whitespace]
    BC --> BD[Decode Block] --> status2{More data?}
    status2 -->|no| C
    status2 -->|yes| BC
    status2 -->|failure|D{{Fail}}
```
(I wanted to use GH's support for mermaid once....I'm aware you know the problem-doamin, we talked enough about this).

heathbm committed:
Address PR feedback: Reuse existing decoding method with whitespace

heathbm commented:
@gfoidl I have updated the PR based on your code. However, despite this code appearing much less complicated and easier to read, I am experiencing some rather unexpected errors.   

Take for instance https://github.com/dotnet/runtime/pull/79334/checks?check_run_id=10282975556.   
The `BasicDecodingWithFinalBlockTrueKnownInputInvalid` test is failing. However, if you run this locally it passes. To get it to fail locally, I ran this test in a loop 100,000 times, and after a few attempts, I get a failure:

````C#
Test has multiple result outcomes
   3 Passed
   2 Failed

Results

    5)   System.Buffers.Text.Tests.Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(inputString: "AQIDBAU", expectedConsumed: 4, expectedWritten: 3) 
      Duration: < 1 ms

      Message: 
Assert.Equal() Failure
Expected: 4
Actual:   8

      Stack Trace: 
Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(String inputString, Int32 expectedConsumed, Int32 expectedWritten) line 273
InvokeStub_Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(Object, Object, IntPtr*)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

    3)   System.Buffers.Text.Tests.Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(inputString: "AQI", expectedConsumed: 0, expectedWritten: 0) 
      Duration: 2 ms

      Message: 
Assert.Equal() Failure
Expected: 0
Actual:   4

      Stack Trace: 
Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(String inputString, Int32 expectedConsumed, Int32 expectedWritten) line 273
InvokeStub_Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(Object, Object, IntPtr*)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

    4)   System.Buffers.Text.Tests.Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(inputString: "AQIDBA", expectedConsumed: 4, expectedWritten: 3) 
      Duration: 6 ms

    2)   System.Buffers.Text.Tests.Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(inputString: "AQ", expectedConsumed: 0, expectedWritten: 0) 
      Duration: 4 ms

    1)   System.Buffers.Text.Tests.Base64DecoderUnitTests.BasicDecodingWithFinalBlockTrueKnownInputInvalid(inputString: "A", expectedConsumed: 0, expectedWritten: 0) 
      Duration: 14 ms

````

The same can be seen in the `BasicDecodingInvalidInputLength` test. This too needs to be run in a loop with multiple attempts:

````C#
  Message: 
Assert.Equal() Failure
Expected: 668136
Actual:   668140

  Stack Trace: 
Base64DecoderUnitTests.BasicDecodingInvalidInputLength() line 55
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) line 59
````

It appears to only be happening in `DecodeFromUtf8` and not `DecodeFromUtf8InPlace`.  
Also, I believe these errors only occur within a built runtime. As I was running these tests in a console environment with the `Base64Decoder.cs` class also in the console project, and I was not able to get these tests to fail there.

I have not looked too much into this at the moment, but perhaps you can spot what it may be already?

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -1125,34 +952,26 @@ private static unsafe void WriteThreeLowOrderBytes(byte* destination, int value)
-                    return i;
-                }
-            }
+                const ulong MagicConstant = 0xC800010000000000UL;
```
At the moment we know how this magic constant is created. But do we know it still next year? I'm rather sure I'll forget it. So please but a comment similar [to this one](https://github.com/gfoidl/Base64/blob/a4968adf3d9b68390797a1290a6ff1e9f61cf7e3/source/gfoidl.Base64/Internal/Operations/CommonOperation.cs#L15-L31) here (at least the link to the _HexConverter_).

gfoidl reviewed (commented):
Looks much better now 👍🏻 Thanks for keeping up with this!

GH doesn't allow me to comment on this place in the code, but can https://github.com/dotnet/runtime/blob/dca7ee6d6bdc28d8d2af0a111731f8584baaba0f/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L133-L138 be removed now? We check for empty spans before, except I miss a special now. Please check this.

My code operates on spans and managed pointers (ref) throughout. This code uses raw-pointers in the core decode methods (only). So one could move the pinning up-front and use raw-pointers throughout.
The advantage is that pinning only happens once, instead on every call to the core-method. But pinning is cheap (especially in contrast to the work done for decoding) and the span-based code reads IMO so much nicer. Thus I'd keep it as is, but it's up to you.

For the test-failures you reported, I'll have a look a bit later.

gfoidl commented:
For the test-failure I believe this is a latent bug present in current _main_, and as this change is on top the already existing building-blocks it unfortunately only manifests here.

Let's take the value from CI `AQI`. So `utf8.Length = 3`. We arrive at https://github.com/dotnet/runtime/blob/dca7ee6d6bdc28d8d2af0a111731f8584baaba0f/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L183-L188 with `srcLength = 0` thus `maxSrcLength = -4`. Then a few steps later we're at https://github.com/dotnet/runtime/blob/dca7ee6d6bdc28d8d2af0a111731f8584baaba0f/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L198-L200 with `(uint)maxScrLength = 4294967292` which is `> int.MaxValue` so the check for the while-loop will always be true, thus entering the loop and calling https://github.com/dotnet/runtime/blob/dca7ee6d6bdc28d8d2af0a111731f8584baaba0f/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L923-L933. As we only have 3 _encodedBytes_ only `t0, t1, t2` are safe to read, whilst `t3` _can_ contain garbage. 
If `t3` happens to have a value that maps to an invalid value, all is good by luck.
If `t3` happens to have a value that maps to an valid value, it's wrong.
Either way that seems to be a 🐛.

Let's continue the journey by assuming invalid data is detected. So we get to https://github.com/dotnet/runtime/blob/dca7ee6d6bdc28d8d2af0a111731f8584baaba0f/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L345-L346 and the following loop that tries to fill that buffer -- here with 3 valid bytes and then passes this to the aforementioned place, where incorrectly the 4th byte will be read, which will be stack-space garbage (that is very likely, especially as SkipLocalsInit is used).

I wonder why this didn't show up before.

In contrast in my implementation the scalar decoding-loop is [guarded by `nint`-casts](https://github.com/gfoidl/Base64/blob/a4968adf3d9b68390797a1290a6ff1e9f61cf7e3/source/gfoidl.Base64/Internal/Base64Impl.Decode.cs#L112-L114) to prevent that underflow (`(uint)-4`). It then goes straight [to this check](https://github.com/gfoidl/Base64/blob/a4968adf3d9b68390797a1290a6ff1e9f61cf7e3/source/gfoidl.Base64/Internal/Base64Impl.Decode.cs#L136-L138) and exits.

If my investigation is correct, then I think this should be handled separated from this PR (as it's a unrelated bug), that might be backported to .NET 6 (LTS).
I'll do some more tests on this -- to get a reliable repro which can be used for (new) tests too.

gfoidl commented:
Created https://github.com/dotnet/runtime/pull/79952 to fix that 🐛.
I suggest you wait until that PR is merged to _main_, then you merge _main_ into your branch to pick the fix up. In the meantime: enjoy your holidays 😃.

heathbm committed:
Merge branch 'main' into add-base64-is-valid

heathbm committed:
Address PR feedback: Remove redundant empty buffer check

heathbm committed:
Address PR Feedback: Add missing magic constant comment

heathbm commented:
Thanks for finding that bug, I've merged in the fix now.

> enjoy your holidays 😃.   

Thanks! I hope you enjoyed yours too! :)

jeffhandley commented:
@heathbm The CI results are still showing a test that is now failing. Per your [previous comment](https://github.com/dotnet/runtime/pull/79334#issuecomment-1341258666), I think this is expected.

https://github.com/dotnet/runtime/pull/79334/checks?check_run_id=10506020439

If that is indeed expected, please update the test to reflect the change in behavior. Would we need to consider that to be a breaking change?

```
System.Security.Cryptography.Tests.Base64TransformsTests.ValidateWhitespace(expected: \"fo\", data: \"\\tZ\\tm8=\\r\\n\")
Exception Message
Assert.Throws() Failure
Expected: typeof(System.FormatException)
Actual:   (No exception was thrown)
Stack Trace
   at System.Security.Cryptography.Tests.Base64TransformsTests.ValidateWhitespace(String expected, String data) in /_/src/libraries/System.Security.Cryptography/tests/Base64TransformsTests.cs:line 304
   at System.Object.InvokeStub_Base64TransformsTests.ValidateWhitespace(Object , Object , IntPtr* )
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs:line 55
```

@gfoidl Since you've been exceptionally thorough with your review and guidance thus far, we'd appreciate your re-review. We can then have someone from the @dotnet/area-system-memory crew review as well once you've approved.

Thanks, folks!

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -34,15 +34,102 @@ public static partial class Base64
+                {
+                    // First char isn't whitespace, but we didn't consume anything,
+                    // thus the input may have whitespace anywhere in between. So fall back to block-wise decoding.
+                    status = DecodeWithWhitespaceBlockwise(utf8, bytes, out bytesConsumed, out bytesWritten, isFinalBlock);
+                    break;
```
Is the `break` here "better" than 
```suggestion
                    return DecodeWithWhitespaceBlockwise(utf8, bytes, out bytesConsumed, out bytesWritten, isFinalBlock);
```
?
If codegen is the same, then I'd prefer `return` as it's easiert to read / follow (exit method here).

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -225,48 +312,135 @@ public static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Spa
+                        if (!IsWhitespace(utf8[i]))
+                        {
+                            // Revert previous dest increment, since an invalid state followed.
+                            bytesConsumed -= localConsumed;
+                            bytesWritten -= localWritten;
```
IMO this is not correct, as whitespace should be treated as consumed, see https://github.com/gfoidl/Base64/pull/184#issuecomment-1363806704

heathbm replied:
Whitespace is being counted, this is when it is not whitespace:   
(https://github.com/dotnet/runtime/pull/79334/files/fb9e8de707c69361e286bf11c401e8d1c7718a86#diff-db463201901c2d83d2b563871ae11fafee9d5afe94e4d014b77212996b25f770R402)

```C#
if (!IsWhitespace(utf8[i]))
{
    // Revert previous dest increment, since an invalid state followed.
    bytesConsumed -= localConsumed;
    bytesWritten -= localWritten;

    return OperationStatus.InvalidData;
}
```

This was to allow existing tests to pass to maintain existing behavior. If the decrements are removed, DecodingInvalidBytesPadding fails.   https://github.com/dotnet/runtime/blob/8da03fa38a8e690af7443702e42042091412a130/src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs#L413

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -350,6 +524,73 @@ public static unsafe OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, ou
+                {
+                    // Padding has already been processed, a new valid block cannot be processed.
+                    // Revert previous dest increment, since an invalid state followed.
+                    localDestIndex -= localBytesWritten;
+                    status = OperationStatus.InvalidData;
```
Same question about correctness?

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,201 @@
+            return true;
+        }
+
+        private static readonly IndexOfAnyValues<byte> validBase64Bytes = IndexOfAnyValues.Create(EncodingMap);
+
+        private static readonly IndexOfAnyValues<char> validBase64Chars = IndexOfAnyValues.Create("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/");
```
Per naming-convections (need to update the references too):
```suggestion
        private static readonly IndexOfAnyValues<byte> s_validBase64Bytes = IndexOfAnyValues.Create(EncodingMap);

        private static readonly IndexOfAnyValues<char> s_validBase64Chars = IndexOfAnyValues.Create("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/");
```

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,201 @@
+                    if (paddingCount == 0)
+                    {
+                        indexOfPaddingInvalidOrWhitespace = base64Text
+                            .Slice(indexOfPaddingInvalidOrWhitespace + 1, base64Text.Length - indexOfPaddingInvalidOrWhitespace - 1)
```
```suggestion
                            .Slice(indexOfPaddingInvalidOrWhitespace + 1)
```
slice to the end?

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,201 @@
+                    if (paddingCount == 0)
+                    {
+                        indexOfPaddingInvalidOrWhitespace = base64TextUtf8
+                            .Slice(indexOfPaddingInvalidOrWhitespace + 1, base64TextUtf8.Length - indexOfPaddingInvalidOrWhitespace - 1)
```
```suggestion
                            .Slice(indexOfPaddingInvalidOrWhitespace + 1)
```
Same here?

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <param name="decodedLength">The maximum length (in bytes) if you were to decode the base 64 encoded text <paramref name="base64TextUtf8"/> within a byte span.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static bool IsValid(ReadOnlySpan<byte> base64TextUtf8, out int decodedLength)
```
The implementation of `byte` and `char` is the same. This duplication can be avoided by using static abstract interface members (with internal / private struct types provide the actual implementation where the difference between byte and char exists -- struct to allow generic specialization, thus better codegen and may inline).

gfoidl commented:
@heathbm do you plan to continue this PR?

heathbm commented:
> @heathbm do you plan to continue this PR?

Apologies for the radio silence, I have not had free time lately. I do plan on continuing.
Also, I recall looking into this failing test (https://github.com/dotnet/runtime/pull/79334#issuecomment-1341258666), I did not see an immediately straight forward way to elegantly resolve this. Perhaps I missed it or perhaps this is a last hurdle.

Note: If someone else has a complete and satisfactory PR earlier, I will be happy to close this one.

@gfoidl Thanks for your continued help and guidance.

heathbm committed:
Address PR Feedback: Avoid validation logic duplication

heathbm committed:
Merge branch 'main' into add-base64-is-valid

gfoidl commented:
Let's bring this PR to an end -- someone from the team please re-open.

heathbm committed:
Throw Base64FormatException when whitespace should not be ignored

This fixes a failing test.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,153 @@
+
+        internal interface IBase64Validatable<T>
+        {
+            static abstract int IndexOfAnyExcept(ReadOnlySpan<T> span);
+            static abstract bool IsWhitespace(T value);
+            static abstract bool IsEncodingPad(T value);
```
I really ❤️ this possibility in the language.

gfoidl reviewed (commented):
Some nits about naming.

My only open question is about https://github.com/dotnet/runtime/pull/79334#discussion_r1091677541
A test with `AQ== ` should have `consumed = 5`. If this test (and all others) pass, then I'm happy with this PR.

When existing tests with invalid input start to fail, then these tests should be validated for the correct behavior.

heathbm committed:
Adress PR feedback: Improve naming of Base64Validator.cs internals

heathbm committed:
Adress PR feedback: Add test to demonstrate extra whitespace is counted

heathbm commented:
> A test with AQ==  should have consumed = 5
@gfoidl Here is a test that covers some extra whitespace at the end scenarios: https://github.com/dotnet/runtime/pull/79334/commits/5848058e6327ed5cc139192d93ac8854f5d28844

gfoidl reviewed (approved):
Only one question regarding tests left + 2 nits, otherwise LGTM.

heathbm commented on src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs:
```diff
@@ -667,5 +672,96 @@ public void DecodeInPlaceInvalidBytesPadding()
+        }
+
+        [Theory]
+        [InlineData("AQ==", 4, 1)]
+        [InlineData("AQ== ", 5, 1)]
```
I covered some of these cases here: https://github.com/dotnet/runtime/pull/79334/files#diff-6b4fec85572fdbcc9b03c81fe66cf8bf31cbb5620235abce7dab3c6dc034d862R11 
It's a lighter version of the gist, as I was not sure if a test with that many loops/asserts would be acceptable in the pipeline.

heathbm commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -34,15 +34,101 @@ public static partial class Base64
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetMaxDecodedFromUtf8Length(int length)
+        {
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length);
```
This method just got moved around, it is untouched from the original: https://github.com/dotnet/runtime/blob/f52e277f54a9413d2f0bd42b8b957c8e4fd263ad/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#LL238C24-L238C24

gfoidl replied:
Yep, I know. It's just as this PR touches that code already that piece can be re-visited to take the new throw-helpers.
I'm not sure what's the general guidance here in runtime -- so whether keep the current implementation or the new possibilities. (for the latter it should be validated if it this method still inlines)

I'll leave that to the maintainers though.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs:
```diff
@@ -0,0 +1,151 @@
+                    // If no padding is found, slice and use IndexOfAnyExcept to look for the next invalid char.
+                    if (paddingCount == 0)
+                    {
+                        ReadOnlySpan<T> slicedSpan = base64Text.Slice(indexOfPaddingInvalidOrWhitespace + 1);
+                        indexOfPaddingInvalidOrWhitespace =
+                            TBase64Validatable.IndexOfAnyExcept(slicedSpan)
+                            + indexOfPaddingInvalidOrWhitespace + 1; // Add current index offset.
```
Looks like this method is still missing test coverage. I'm pretty sure input like `" a"` would go into an infinite loop here:
- `indexOfPaddingInvalidOrWhitespace` gets set to 0 because of the initial whitespace
- `paddingCount == 0`, so we get to this branch
- `slicedSpan` = `"a"`
- `IndexOfAnyExcept` returns -1, so you set the index to `-1 + 0 + 1`
- the index is 0 again ... repeat the loop

It may be simpler to write this along the lines of
```c#
int length = 0;

while (true)
{
    int index = IndexOfAnyExcept(base64Text);
    if ((uint)index >= (uint)base64Text.Length)
    {
        length += base64Text.Length;
        break;
    }

    length += index;

    T charToValidate = base64Text[index];
    base64Text = base64Text.Slice(index + 1);

    if (IsWhitespace(charToValidate))
    {
        continue;
    }

    if (!IsEncodingPad(charToValidate))
    {
        return false;
    }

    int paddingCount = 1;
    foreach (T char in base64Text)
    {
        if (IsEncodingPad(char))
        {
            paddingCount++;
        }
        else if (!IsWhitespace(char))
        {
            return false;
        }
    }

    if (paddingCount > 2)
    {
        return false;
    }

    length += paddingCount;
    break;
}

if (length % 4 != 0)
{
    return false;
}
```

heathbm replied:
Apologies I should not have missed that! Thank you for the suggestion, it is much more elegant. I have just made a few slight changes:
- Moved the padding validation outside of the while loop.  
- Return false as soon as more than 2 padding chars detected, instead of counting all then determining if more than 2.  

Please let me know if you preferred it the other way.

heathbm committed:
Address PR feedback: avoid bound-check

heathbm committed:
Address PR feedback: Base64.IsValid: Return when no more invalid chars

heathbm committed:
Address PR feedback: Refactor Bas64.IsValid method

tarekgh commented:
@heathbm could you please resolve the merge conflict?

tarekgh commented:
@heathbm could you please resolve the merge conflict?

tarekgh commented:
@dotnet/area-system-memory can one of you review this change to unblock the PR and merge it?

tarekgh commented:
@heathbm could you please resolve the merge conflict?

tarekgh commented:
@heathbm any news?

heathbm commented:
@tarekgh I would love to work on this, however I have no free cycles. I did check, and it looks like Base64Transforms.cs was refactored. The methods I modified no longer exist. This would require a different approach from what I initially proposed.

tarekgh commented:
Thanks @heathbm for your feedback. I am closing this PR as it became stale and needs a fully different fix.
